develop/reference_workarounds.rst

```diff
@@ -128,3 +128,201 @@ There is another trick though. We can pull the relevant information to the coord
+
+When two distributed tables are :ref:`colocated <colocation>`, then we can create a trigger to modify one based on changes in the other. The idea, once again, is to create triggers on the placements, but the trigger must be between pairs of placements that are themselves colocated. For this Citus has a special helper function ``run_command_on_colocated_placements``.
+
+Suppose that for every value inserted into ``little_vals`` we want to insert one twice as big into ``big_vals``.
+
+.. code-block:: postgresql
+
+  CREATE TABLE little_vals (key int, val int);
+  CREATE TABLE big_vals    (key int, val int);
+  SELECT create_distributed_table('little_vals', 'key');
+  SELECT create_distributed_table('big_vals',    'key');
+
+  -- This trigger function takes the destination placement as an argument
+
+  SELECT run_command_on_workers($cmd$
+    CREATE OR REPLACE FUNCTION embiggen() RETURNS TRIGGER AS $$
+      BEGIN
+        IF (TG_OP = 'INSERT') THEN
+          EXECUTE format(
+            'INSERT INTO %s (key, val) SELECT ($1).key, ($1).val*2;',
+            TG_ARGV[0]
+          ) USING NEW;
+        END IF;
+        RETURN NULL;
+      END;
+    $$ LANGUAGE plpgsql;
+  $cmd$);
+
+  -- Next we relate the co-located tables by the trigger function
+  -- on each co-located placement
+
+  SELECT run_command_on_colocated_placements(
+    'little_vals',
+    'big_vals',
+    $cmd$
+      CREATE TRIGGER after_insert AFTER INSERT ON %s
+        FOR EACH ROW EXECUTE PROCEDURE embiggen(%s)
+    $cmd$
+  );
+
+Then to test it:
+
+.. code-block:: postgresql
+
+  INSERT INTO little_vals VALUES (1, 42), (2, 101);
+  TABLE big_vals;
+
+::
+
+  ┌─────┬─────┐
+  │ key │ val │
+  ├─────┼─────┤
+  │   1 │  84 │
+  │   2 │ 202 │
+  └─────┴─────┘
+
+**Trigger between reference tables.**
+
+Reference tables are simpler than distributed tables in that they have exactly one shard which is replicated across all workers. To relate reference tables with a trigger, we simply need to create the trigger for the shard on all workers.
+
+Suppose we want to record every change of ``insert_target`` in ``audit_table``, both of which are reference tables.
+
+.. code-block:: postgresql
+
+  -- create the reference tables
+
+  CREATE TABLE insert_target (
+    value text
+  );
+  CREATE TABLE audit_table(
+    stamp timestamp,
+    value text
+  );
+  SELECT create_reference_table('insert_target');
+  SELECT create_reference_table('audit_table');
+
+  -- now find the shard id for each table
+
+  SELECT logicalrelid, shardid
+  FROM pg_dist_shard
+  WHERE logicalrelid IN
+    ('insert_target'::regclass, 'audit_table'::regclass);
+
+Include the shard ids (written below as "xxxxxx" and "yyyyyy") in custom queries:
+
+.. code-block:: postgresql
+
+  SELECT run_command_on_workers($cmd$
+    CREATE OR REPLACE FUNCTION process_audit() RETURNS TRIGGER AS $$
+      BEGIN
+        INSERT INTO audit_table_xxxxxx(stamp,value) VALUES (now(),'value');
+        RETURN NEW;
```

Comment:
Note that this is unsafe in many cases because it by-passes the write locks on the reference table. Therefore, concurrent writes to audit_table may be applied in a different order on different replicas which would cause replicas to diverge.

Reply:
It's also unsafe for replication factor > 1

Reply:
Is it because of the now()? @begriffs In a customer example, we had a timestamp on the `insert_target` table, and used that value in the `audit_table` using `NEW.`. That one gets evaluated on the master, so the value remains the same across all placements.

Reply:
@marcocitus is there a better workaround, or is there no safe way to do triggers between reference tables?

Reply:
It's because the order in which concurrent update/delete/upsert commands are applied to replica's is not guaranteed. Technically, an insert-only example is safe because insertion order doesn't really matter, but small deviations can cause it not to be. There is no workaround right now. In MX you will be able to update another reference table from a trigger on a single shard placement. That has some other downsides (it creates hierarchies of cached connections), but is safe.